Feat: Valkey support for Elasticache Ltang/horde/valkey

ltang-aws opened the pull request:
**Issue number:**
closes #509

## Summary
Implemented Valkey as a supported elasticache engine type when deploying Horde. It is now the default engine.

### Changes

> Please provide a summary of what's being changed

Valkey can be deployed as a replication group instead of as a Redis cluster. Generalized some variable names to be non-Redis exclusive. Added appropriate conditionals to IAM permissions

### User experience

> Please share what the user experience looks like before and after this change

From the main.tf, user can change engine:

elasticache_engine                = "valkey"
to 
elasticache_engine                = "redis"

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [Yes] I have performed a self-review of this change
* [Yes] Changes have been tested
* [See Below] Changes are documented

### Documentation
Henry would like this change to be documented on https://aws-games.github.io/cloud-game-development-toolkit/latest/modules/unreal/horde/unreal-engine-horde.html

Right now information on Horde is split under Unreal Engine Horde and Examples. This needs to be further organized and may affect the placement of the valkey documentation


### Draft of documentation

The Cloud Game Development Toolkit supports the deployment of Horde with Redis as a cluster or Valkey as a replication group. To change engines, change the elasticache_engine parameter to be either "redis' or "valkey" in modules/horde/examples/complete/main.tf



<details>
<summary>Is this a breaking change?</summary>

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created might not be successful.

ltang-aws commented:
@henrykie @gabebatista Hey there, could I get some eyes on this to see if everything looks good?

gabebatista commented on modules/unreal/horde/variables.tf:
```diff
@@ -349,9 +349,25 @@ variable "docdb_storage_encrypted" {
+  description = "The engine to use for ElastiCache (redis or valkey)"
   type        = string
-  description = "The redis connection configuration that Horde should use."
+  default     = "redis"
```
Why default to Redis? I'm not saying you're wrong, I just want to understand how they compare to each other.

henrykie replied:
I am split on this. From the Game DevOps slack there have been a few people express that Valkey does not seem to behave correctly in Horde... And now that Redis is again open source (@jorisdon ?) we aren't as concerned about the licensing differences. @kylesomers what do you think?

kylesomers replied:
Licensing isn't the motivating factor, its the cost difference between the two engine types per Elasticache pricing. Valkey offers 20% cheaper node price per hour for t4g.micro (this module's default node type). But if there's reported bugs using Valkey, let's identify what those are and determine if this needs further testing before releasing, or at least document these known issues.

gabebatista reviewed (commented):
Left some minor comments but overall this is solid. Great job!

gabebatista commented:
If you are ready to receive feedback on this, convert it into a regular PR instead of a draft

henrykie commented on modules/unreal/horde/elasticache.tf:
```diff
@@ -1,22 +1,38 @@
+  engine                     = "valkey"
+  engine_version             = "7.2"
+  replication_group_id       = "${var.name}-elasticache-valkey-rep-grp"
+  description                = "valkey for horde"
```
Valkey for Unreal Engine Horde

henrykie commented on modules/unreal/horde/local.tf:
```diff
@@ -14,13 +14,18 @@ locals {
   elasticache_redis_parameter_group_name = "default.redis7"
+  elasticache_redis_connection_strings   = var.elasticache_engine == "redis" ? [for node in aws_elasticache_cluster.horde[0].cache_nodes : "${node.address}:${node.port}"] : null
+
+  elasticache_valkey_engine_version       = "7.2"
```
We are not actually using this local. The replication group resource creation ahs 7.2 hardcoded. Let's create a variable: `elasticache_valkey_engine_version` that defaults to 7.2. Same thing for `elasticache_redis_engine_version` and `elasticache_redis_parameter_group_name`. These provide overrides for us to test new versions of the engine and the default parameter group. Remove the locals here.

henrykie commented on modules/unreal/horde/variables.tf:
```diff
@@ -349,9 +349,25 @@ variable "docdb_storage_encrypted" {
+
+variable "elasticache_cluster_count" {
+  type        = number
+  description = "Number of cache cluster to provision in the Elasticache valkey cluster."
```
Here we say "Elasticache valkey cluster" but this is used in both the redis and valkey paths, correct?

ltang-aws commented:
@henrykie @gabebatista I cleaned up the locals and changed them to variables. Keeping the default engine as Redis for now since Kyle mentioned the potential instability of Valkey

henrykie commented:
Please update this branch so it is up-to-date with main. You can do so locally with a rebase, or click the arrow next to the "Update branch" button and do an "Update with rebase."

henrykie commented on modules/unreal/horde/examples/complete/main.tf:
```diff
@@ -17,7 +17,7 @@ module "unreal_engine_horde" {
   github_credentials_secret_arn     = var.github_credentials_secret_arn
   tags                              = local.tags
-
+  elasticache_engine                = "valkey"
```
This is an example, so Kenneth is setting the `elasticache_engine` variable to `valkey` - does that make sense?

ayatanb reviewed (approved):
LGTM yay!
